packages/wallet/src/TxBuilder/types.ts

```diff
@@ -0,0 +1,126 @@
+import { CustomError } from 'ts-custom-error';
+import { DeepPartial } from '@cardano-sdk/util';
+import { InputSelectionError } from '@cardano-sdk/cip2';
+
+import { OutputValidation } from '../types';
+import { SignTransactionOptions } from '../KeyManagement';
+
+// Does not transform TokenMap into Partial<TokenMap> because it would also make it's methods and map types optional
+// and that forces users to do many checks before being able to use any of it's properties and methods
+export type PartialTxOut = DeepPartial<Cardano.TxOut, Cardano.TokenMap>;
+
+export enum TxOutputFailure {
+  MinimumCoin = 'Minimum coin not met',
+  TokenBundleSizeExceedsLimit = 'Token Bundle Exceeds Limit',
+  MissingRequiredFields = 'Mandatory fields address or coin are missing'
+}
+
+export class OutputValidationMissingRequiredError extends CustomError {
+  public constructor(public txOut: PartialTxOut) {
+    super(TxOutputFailure.MissingRequiredFields);
+  }
+}
+
+export class OutputValidationMinimumCoinError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.MinimumCoin);
+  }
+}
+
+export class OutputValidationTokenBundleSizeError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.TokenBundleSizeExceedsLimit);
+  }
+}
+
+export class IncompatibleWalletError extends CustomError {}
+
+export type TxOutValidationError =
+  | OutputValidationMissingRequiredError
+  | OutputValidationMinimumCoinError
+  | OutputValidationTokenBundleSizeError;
+export type TxBodyValidationError = TxOutValidationError | InputSelectionError;
+
+export type Valid<TValid> = TValid & {
+  isValid: true;
+};
+
+export interface Invalid<TError> {
+  isValid: false;
+  errors: TError[];
+}
+
+export interface ValidTxOutData {
+  readonly txOut: Cardano.TxOut;
+}
+
+export type ValidTxOut = Valid<ValidTxOutData>;
+export type InvalidTxOut = Invalid<TxOutValidationError>;
+export type MaybeValidTxOut = ValidTxOut | InvalidTxOut;
+
+export interface OutputBuilder {
+  partialOutput: PartialTxOut;
+  value(value: Cardano.Value): OutputBuilder;
+  coin(coin: Cardano.Lovelace): OutputBuilder;
+  assets(assets: Cardano.TokenMap): OutputBuilder;
+  /**
+   * @param quantity To remove an asset, set quantity to 0
+   */
+  asset(assetId: Cardano.AssetId, quantity: bigint): OutputBuilder;
+  address(address: Cardano.Address): OutputBuilder;
+  datum(datum: Cardano.util.Hash32ByteBase16): OutputBuilder;
+  build(): Promise<MaybeValidTxOut>;
+}
+
+export interface SignedTx {
+  readonly tx: Cardano.NewTxAlonzo;
+  submit(): Promise<void>;
+}
+
+export interface ValidTxBody {
+  readonly body: Cardano.NewTxBodyAlonzo;
+  readonly auxiliaryData?: Cardano.AuxiliaryData;
+  readonly extraWitness?: Partial<Cardano.Witness>;
+
+  sign(props?: SignTransactionOptions): Promise<SignedTx>;
```

Comment:
Options should come from previous step. `initializeTx` has to know the exact number of signatures that the transaction is going to have, so there's no point of having this argument, as it should also be supplied when building the tx.

```suggestion
  sign(): Promise<SignedTx>;
```

Reply:
[53ef574](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/53ef574ab31271196145f491bc145e55c4774848) & [c4037b8](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/c4037b828fc9dc3b4b1fad0446101aaa8f63f4cf)